File: modules/web-console/backend/app/browsersHandler.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

'use strict';

const _ = require('lodash');
const socketio = require('socket.io');

// Fire me up!

/**
 * Module interaction with browsers.
 */
module.exports = {
    implements: 'browsers-handler',
    inject: ['configure', 'errors', 'mongo'],
    factory: (configure, errors, mongo) => {
        class BrowserSockets {
            constructor() {
                this.sockets = new Map();
            }

            /**
             * @param {Socket} sock
             */
            add(sock) {
                const key = sock.request.user._id.toString();

                if (this.sockets.has(key))
                    this.sockets.get(key).push(sock);
                else
                    this.sockets.set(key, [sock]);

                return this.sockets.get(sock.request.user);
            }

            /**
             * @param {Socket} sock
             */
            remove(sock) {
                const key = sock.request.user._id.toString();

                const sockets = this.sockets.get(key);

                _.pull(sockets, sock);

                return sockets;
            }

            get(account) {
                const key = account._id.toString();

                let sockets = this.sockets.get(key);

                if (_.isEmpty(sockets))
                    this.sockets.set(key, sockets = []);

                return sockets;
            }
        }

        class BrowsersHandler {
            /**
             * @constructor
             */
            constructor() {
                /**
                 * Connected browsers.
                 * @type {BrowserSockets}
                 */
                this._browserSockets = new BrowserSockets();

                /**
                 * Registered Visor task.
                 * @type {Map}
                 */
                this._visorTasks = new Map();
            }

            /**
             * @param {Error} err
             * @return {{code: number, message: *}}
             */
            errorTransformer(err) {
                return {
                    code: err.code || 1,
                    message: err.message || err
                };
            }

            /**
             * @param {String} account
             * @param {Array.<Socket>} [socks]
             */
            agentStats(account, socks = this._browserSockets.get(account)) {
                return this._agentHnd.agents(account)
                    .then((agentSocks) => {
                        const stat = _.reduce(agentSocks, (acc, agentSock) => {
                            acc.count += 1;
                            acc.hasDemo = acc.hasDemo || _.get(agentSock, 'demo.enabled');

                            if (agentSock.cluster)
                                acc.clusters.push(agentSock.cluster);

                            return acc;
                        }, {count: 0, hasDemo: false, clusters: []});

                        stat.clusters = _.uniqWith(stat.clusters, _.isEqual);

                        return stat;
                    })
                    .catch(() => ({count: 0, hasDemo: false, clusters: []}))
                    .then((stat) => _.forEach(socks, (sock) => sock.emit('agents:stat', stat)));
            }

            clusterChanged(account, cluster) {
                const socks = this._browserSockets.get(account);

                _.forEach(socks, (sock) => {
                    if (sock)
                        sock.emit('cluster:changed', cluster);
                    else
                        console.log(`Fount closed socket [account=${account}, cluster=${cluster}]`);
                });
            }

            pushInitialData(sock) {
                // Send initial data.
            }

            emitNotification(sock) {
                sock.emit('user:notifications', this.notification);
            }

            /**
             * @param {String} notification Notification message.
             */
            updateNotification(notification) {
                this.notification = notification;

                for (const socks of this._browserSockets.sockets.values()) {
                    for (const sock of socks)
                        this.emitNotification(sock);
                }
            }

            executeOnAgent(account, demo, event, ...args) {
                const cb = _.last(args);

                return this._agentHnd.agent(account, demo)
                    .then((agentSock) => agentSock.emitEvent(event, ..._.dropRight(args)))
                    .then((res) => cb(null, res))
                    .catch((err) => cb(this.errorTransformer(err)));
            }

            agentListeners(sock) {
                const demo = sock.request._query.IgniteDemoMode === 'true';
                const account = () => sock.request.user;

                // Return available drivers to browser.
                sock.on('schemaImport:drivers', (...args) => {
                    this.executeOnAgent(account(), demo, 'schemaImport:drivers', ...args);
                });

                // Return schemas from database to browser.
                sock.on('schemaImport:schemas', (...args) => {
                    this.executeOnAgent(account(), demo, 'schemaImport:schemas', ...args);
                });

                // Return tables from database to browser.
                sock.on('schemaImport:metadata', (...args) => {
                    this.executeOnAgent(account(), demo, 'schemaImport:metadata', ...args);
                });
            }

            /**
             * @param {Promise.<AgentSocket>} agent
             * @param {Boolean} demo
             * @param {{sessionId: String}|{'login': String, 'password': String}} credentials
             * @param {Object.<String, String>} params
             * @return {Promise.<T>}
             */
            executeOnNode(agent, token, demo, credentials, params) {
                return agent
                    .then((agentSock) => agentSock.emitEvent('node:rest',
                        {uri: 'ignite', token, demo, params: _.merge({}, credentials, params)}));
            }

            registerVisorTask(taskId, taskCls, ...argCls) {
                this._visorTasks.set(taskId, {
                    taskCls,
                    argCls
                });
            }

            nodeListeners(sock) {
                // Return command result from grid to browser.
                sock.on('node:rest', (arg, cb) => {
                    const {clusterId, params, credentials} = arg || {};

                    if (!_.isFunction(cb))
                        cb = console.log;

                    const demo = _.get(sock, 'request._query.IgniteDemoMode') === 'true';

                    if ((_.isNil(clusterId) && !demo) || _.isNil(params)) {
                        console.log('Received invalid message: "node:rest" on socket:', JSON.stringify(sock.handshake));

                        return cb('Invalid format of message: "node:rest"');
                    }

                    const agent = this._agentHnd.agent(sock.request.user, demo, clusterId);

                    const token = sock.request.user.token;

                    this.executeOnNode(agent, token, demo, credentials, params)
                        .then((data) => cb(null, data))
                        .catch((err) => cb(this.errorTransformer(err)));
                });

                const internalVisor = (postfix) => `org.apache.ignite.internal.visor.${postfix}`;

                this.registerVisorTask('querySql', internalVisor('query.VisorQueryTask'), internalVisor('query.VisorQueryArg'));
                this.registerVisorTask('querySqlV2', internalVisor('query.VisorQueryTask'), internalVisor('query.VisorQueryArgV2'));
                this.registerVisorTask('querySqlV3', internalVisor('query.VisorQueryTask'), internalVisor('query.VisorQueryArgV3'));
                this.registerVisorTask('querySqlX2', internalVisor('query.VisorQueryTask'), internalVisor('query.VisorQueryTaskArg'));

                this.registerVisorTask('queryScanX2', internalVisor('query.VisorScanQueryTask'), internalVisor('query.VisorScanQueryTaskArg'));

                this.registerVisorTask('queryFetch', internalVisor('query.VisorQueryNextPageTask'), 'org.apache.ignite.lang.IgniteBiTuple', 'java.lang.String', 'java.lang.Integer');
                this.registerVisorTask('queryFetchX2', internalVisor('query.VisorQueryNextPageTask'), internalVisor('query.VisorQueryNextPageTaskArg'));

                this.registerVisorTask('queryFetchFirstPage', internalVisor('query.VisorQueryFetchFirstPageTask'), internalVisor('query.VisorQueryNextPageTaskArg'));
                this.registerVisorTask('queryPing', internalVisor('query.VisorQueryPingTask'), internalVisor('query.VisorQueryNextPageTaskArg'));

                this.registerVisorTask('queryClose', internalVisor('query.VisorQueryCleanupTask'), 'java.util.Map', 'java.util.UUID', 'java.util.Set');
                this.registerVisorTask('queryCloseX2', internalVisor('query.VisorQueryCleanupTask'), internalVisor('query.VisorQueryCleanupTaskArg'));

                this.registerVisorTask('toggleClusterState', internalVisor('misc.VisorChangeGridActiveStateTask'), internalVisor('misc.VisorChangeGridActiveStateTaskArg'));

                this.registerVisorTask('cacheNamesCollectorTask', internalVisor('cache.VisorCacheNamesCollectorTask'), 'java.lang.Void');

                this.registerVisorTask('cacheNodesTask', internalVisor('cache.VisorCacheNodesTask'), 'java.lang.String');
                this.registerVisorTask('cacheNodesTaskX2', internalVisor('cache.VisorCacheNodesTask'), internalVisor('cache.VisorCacheNodesTaskArg'));

                // Return command result from grid to browser.
                sock.on('node:visor', (arg, cb) => {
                    const {clusterId, params, credentials} = arg || {};

                    if (!_.isFunction(cb))
                        cb = console.log;

                    const demo = _.get(sock, 'request._query.IgniteDemoMode') === 'true';

                    if ((_.isNil(clusterId) && !demo) || _.isNil(params)) {
                        console.log('Received invalid message: "node:visor" on socket:', JSON.stringify(sock.handshake));

                        return cb('Invalid format of message: "node:visor"');
                    }

                    const {taskId, nids, args = []} = params;

                    const desc = this._visorTasks.get(taskId);

                    if (_.isNil(desc))
                        return cb(this.errorTransformer(new errors.IllegalArgumentException(`Failed to find Visor task for id: ${taskId}`)));

                    const exeParams = {
                        cmd: 'exe',
                        name: 'org.apache.ignite.internal.visor.compute.VisorGatewayTask',
                        p1: nids,
                        p2: desc.taskCls
                    };

                    _.forEach(_.concat(desc.argCls, args), (param, idx) => { exeParams[`p${idx + 3}`] = param; });

                    const agent = this._agentHnd.agent(sock.request.user, demo, clusterId);

                    const token = sock.request.user.token;

                    this.executeOnNode(agent, token, demo, credentials, exeParams)
                        .then((data) => {
                            if (data.finished && !data.zipped)
                                return cb(null, data.result);

                            return cb(null, data);
                        })
                        .catch((err) => cb(this.errorTransformer(err)));
                });
            }

            /**
             *
             * @param server
             * @param {AgentsHandler} agentHnd
             */
            attach(server, agentHnd) {
                this._agentHnd = agentHnd;

                if (this.io)
                    throw 'Browser server already started!';

                mongo.Notifications.findOne().sort('-date').exec()
                    .then((notification) => {
                        this.notification = notification;
                    })
                    .then(() => {
                        const io = socketio(server);

                        configure.socketio(io);

                        // Handle browser connect event.
                        io.sockets.on('connection', (sock) => {
                            this._browserSockets.add(sock);

                            // Handle browser disconnect event.
                            sock.on('disconnect', () => {
                                this._browserSockets.remove(sock);
                            });

                            this.agentListeners(sock);
                            this.nodeListeners(sock);

                            this.pushInitialData(sock);
                            this.agentStats(sock.request.user, [sock]);
                            this.emitNotification(sock);
                        });
                    });
            }
        }

        return new BrowsersHandler();
    }
};

File: modules/web-console/web-console-server/src/main/java/org/apache/ignite/console/web/socket/BrowsersHandler.java
/*
 * Copyright 2019 GridGain Systems, Inc. and Contributors.
 *
 * Licensed under the GridGain Community Edition License (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.ignite.console.web.socket;

import java.security.Principal;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.Stream;
import org.apache.ignite.console.dto.Account;
import org.apache.ignite.console.json.JsonArray;
import org.apache.ignite.console.json.JsonObject;
import org.apache.ignite.console.messages.WebConsoleMessageSource;
import org.apache.ignite.console.messages.WebConsoleMessageSourceAccessor;
import org.apache.ignite.console.web.AbstractHandler;
import org.apache.ignite.console.web.model.VisorTaskDescriptor;
import org.apache.ignite.console.websocket.WebSocketRequest;
import org.apache.ignite.console.websocket.WebSocketResponse;
import org.apache.ignite.internal.util.typedef.F;
import org.apache.ignite.lang.IgniteBiTuple;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.web.socket.CloseStatus;
import org.springframework.web.socket.WebSocketSession;

import static org.apache.ignite.console.utils.Utils.fromJson;
import static org.apache.ignite.console.websocket.WebSocketEvents.NODE_REST;
import static org.apache.ignite.console.websocket.WebSocketEvents.NODE_VISOR;
import static org.apache.ignite.console.websocket.WebSocketEvents.SCHEMA_IMPORT_DRIVERS;
import static org.apache.ignite.console.websocket.WebSocketEvents.SCHEMA_IMPORT_METADATA;
import static org.apache.ignite.console.websocket.WebSocketEvents.SCHEMA_IMPORT_SCHEMAS;

/**
 * Browsers web sockets handler.
 */
@Service
public class BrowsersHandler extends AbstractHandler {
    /** */
    private static final Logger log = LoggerFactory.getLogger(BrowsersHandler.class);

    /** */
    private static final String VISOR_IGNITE = "org.apache.ignite.internal.visor.";

    /** */
    private final Map<String, VisorTaskDescriptor> visorTasks = new HashMap<>();

    /** */
    private final WebSocketsManager wsm;

    /** Messages accessor. */
    private WebConsoleMessageSourceAccessor messages = WebConsoleMessageSource.getAccessor();

    /**
     * @param wsm Web sockets manager.
     */
    public BrowsersHandler(WebSocketsManager wsm) {
        this.wsm = wsm;

        registerVisorTasks();
    }

    /**
     * Extract account from session.
     *
     * @param ws Websocket.
     * @return Account.
     */
    protected Account extractAccount(WebSocketSession ws) {
        Principal p = ws.getPrincipal();

        if (p instanceof Authentication) {
            Authentication t = (Authentication)p;

            Object tp = t.getPrincipal();

            if (tp instanceof Account)
                return (Account)tp;
        }

        throw new IllegalStateException(messages.getMessageWithArgs("err.account-cant-be-found-in-ws-session", ws));
    }

    /** {@inheritDoc} */
    @Override public void afterConnectionEstablished(WebSocketSession ws) {
        log.info("Browser session opened [socket=" + ws + "]");

        ws.setTextMessageSizeLimit(10 * 1024 * 1024);

        Account acc = extractAccount(ws);

        wsm.onBrowserConnect(ws, acc.getId());
    }

    /** {@inheritDoc} */
    @Override public void handleEvent(WebSocketSession ws, WebSocketRequest evt) {
        try {
            switch (evt.getEventType()) {
                case SCHEMA_IMPORT_DRIVERS:
                case SCHEMA_IMPORT_SCHEMAS:
                case SCHEMA_IMPORT_METADATA:
                    wsm.sendToFirstAgent(ws, evt.response());

                    break;

                case NODE_REST:
                case NODE_VISOR:
                    JsonObject payload = fromJson(evt.getPayload());

                    String clusterId = payload.getString("clusterId");

                    if (F.isEmpty(clusterId))
                        throw new IllegalStateException(messages.getMessage("err.missing-cluster-id-param"));

                    WebSocketResponse reqEvt = evt.getEventType().equals(NODE_REST) ?
                        evt.response() : evt.withPayload(prepareNodeVisorParams(payload));

                    wsm.sendToNode(ws, clusterId, reqEvt);

                    break;

                default:
                    throw new IllegalStateException(messages.getMessageWithArgs("err.unknown-evt", evt));
            }
        }
        catch (IllegalStateException e) {
            log.warn(e.toString());

            sendError(ws, evt, "Failed to send event to agent: " + evt.getPayload(), e);
        }
        catch (Throwable e) {
            String errMsg = "Failed to send event to agent: " + evt.getPayload();

            log.error(errMsg, e);

            sendError(ws, evt, errMsg, e);
        }
    }

    /** {@inheritDoc} */
    @Override public void afterConnectionClosed(WebSocketSession ws, CloseStatus status) {
        log.info("Browser session closed [socket=" + ws + ", status=" + status + "]");

        wsm.onBrowserConnectionClosed(ws);
    }

    /**
     * @param shortName Class short name.
     * @return Full class name.
     */
    protected String igniteVisor(String shortName) {
        return VISOR_IGNITE + shortName;
    }

    /**
     * @param taskId Task ID.
     * @param taskCls Task class name.
     * @param argCls Arguments classes names.
     */
    protected void registerVisorTask(String taskId, String taskCls, String... argCls) {
        visorTasks.put(taskId, new VisorTaskDescriptor(taskCls, argCls));
    }

    /**
     * Register Visor tasks.
     */
    private void registerVisorTasks() {
        registerVisorTask(
            "querySql",
            igniteVisor("query.VisorQueryTask"),
            igniteVisor("query.VisorQueryArg"));

        registerVisorTask("querySqlV2",
            igniteVisor("query.VisorQueryTask"),
            igniteVisor("query.VisorQueryArgV2"));

        registerVisorTask("querySqlV3",
            igniteVisor("query.VisorQueryTask"),
            igniteVisor("query.VisorQueryArgV3"));

        registerVisorTask("querySqlX2",
            igniteVisor("query.VisorQueryTask"),
            igniteVisor("query.VisorQueryTaskArg"));

        registerVisorTask("queryScanX2",
            igniteVisor("query.VisorScanQueryTask"),
            igniteVisor("query.VisorScanQueryTaskArg"));

        registerVisorTask("queryFetch",
            igniteVisor("query.VisorQueryNextPageTask"),
            IgniteBiTuple.class.getName(), String.class.getName(), Integer.class.getName());

        registerVisorTask("queryFetchX2",
            igniteVisor("query.VisorQueryNextPageTask"),
            igniteVisor("query.VisorQueryNextPageTaskArg"));

        registerVisorTask("queryFetchFirstPage",
            igniteVisor("query.VisorQueryFetchFirstPageTask"),
            igniteVisor("query.VisorQueryNextPageTaskArg"));

        registerVisorTask("queryClose",
            igniteVisor("query.VisorQueryCleanupTask"),
            Map.class.getName(), UUID.class.getName(), Set.class.getName());

        registerVisorTask("queryCloseX2",
            igniteVisor("query.VisorQueryCleanupTask"),
            igniteVisor("query.VisorQueryCleanupTaskArg"));

        registerVisorTask("toggleClusterState",
            igniteVisor("misc.VisorChangeGridActiveStateTask"),
            igniteVisor("misc.VisorChangeGridActiveStateTaskArg"));

        registerVisorTask("cacheNamesCollectorTask",
            igniteVisor("cache.VisorCacheNamesCollectorTask"),
            Void.class.getName());

        registerVisorTask("cacheNodesTask",
            igniteVisor("cache.VisorCacheNodesTask"),
            String.class.getName());

        registerVisorTask("cacheNodesTaskX2",
            igniteVisor("cache.VisorCacheNodesTask"),
            igniteVisor("cache.VisorCacheNodesTaskArg"));
    }

    /**
     * Prepare task event for execution on agent.
     *
     * @param payload Task event.
     */
    private JsonObject prepareNodeVisorParams(JsonObject payload) {
        JsonObject params = payload.getJsonObject("params");

        String taskId = params.getString("taskId");

        if (F.isEmpty(taskId))
            throw new IllegalStateException(messages.getMessageWithArgs("err.not-specified-task-id", payload));

        String nids = params.getString("nids");

        VisorTaskDescriptor desc = visorTasks.get(taskId);

        if (desc == null)
            throw new IllegalStateException(messages.getMessageWithArgs("err.unknown-task", taskId, payload));

        JsonObject exeParams =  new JsonObject()
            .add("cmd", "exe")
            .add("name", "org.apache.ignite.internal.visor.compute.VisorGatewayTask")
            .add("p1", nids)
            .add("p2", desc.getTaskClass());

        AtomicInteger idx = new AtomicInteger(3);

        Arrays.stream(desc.getArgumentsClasses()).forEach(arg ->  exeParams.put("p" + idx.getAndIncrement(), arg));

        JsonArray args = params.getJsonArray("args");

        if (!F.isEmpty(args))
            args.forEach(arg -> exeParams.put("p" + idx.getAndIncrement(), arg));

        Stream.of("user", "password", "sessionToken").forEach(p -> exeParams.add(p, params.get(p)));

        payload.put("params", exeParams);

        return payload;
    }
}
